File: pkg/adapter/http/handlers.go
package http

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"

	"github.com/ModulationAI/openagentio/pkg/bus"
	"github.com/ModulationAI/openagentio/pkg/event"
)

// authenticate runs the configured AuthFunc and returns the derived context.
// Returns (nil, nil, false) when authentication failed and a 401 has already
// been written; callers must short-circuit. Returns (ac, true) when allowed.
func (a *Adapter) authenticate(w http.ResponseWriter, r *http.Request) (*AuthContext, bool) {
	if a.auth == nil {
		return nil, true
	}
	ac, err := a.auth(r)
	if err != nil {
		writeErrorJSON(w, http.StatusUnauthorized, event.CodeAuthFailure, err.Error())
		return nil, false
	}
	return ac, true
}

// handleInvoke implements POST /v1/agents/{target}/invoke.
func (a *Adapter) handleInvoke(w http.ResponseWriter, r *http.Request) {
	ac, ok := a.authenticate(w, r)
	if !ok {
		return
	}
	target := r.PathValue("target")
	if target == "" {
		writeErrorJSON(w, http.StatusBadRequest, event.CodeInvalidRequest, "missing target")
		return
	}
	env, err := readEnvelope(r, "", ac)
	if err != nil {
		writeErrorJSON(w, http.StatusBadRequest, event.CodeInvalidRequest, err.Error())
		return
	}

	ctx := r.Context()
	if a.timeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, a.timeout)
		defer cancel()
	}

	resp, err := a.bus.Invoke(ctx, target, env, bus.WithTimeout(a.timeout))
	if err != nil {
		writeBusError(w, err)
		return
	}

	if resp.EventType == event.ResponseError {
		writeEnvelopeError(w, resp)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	if len(resp.Payload) == 0 {
		w.WriteHeader(http.StatusNoContent)
		return
	}
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write(resp.Payload)
}

// handleStream implements POST /v1/agents/{target}/stream and emits SSE.
func (a *Adapter) handleStream(w http.ResponseWriter, r *http.Request) {
	ac, ok := a.authenticate(w, r)
	if !ok {
		return
	}
	target := r.PathValue("target")
	if target == "" {
		writeErrorJSON(w, http.StatusBadRequest, event.CodeInvalidRequest, "missing target")
		return
	}
	env, err := readEnvelope(r, "", ac)
	if err != nil {
		writeErrorJSON(w, http.StatusBadRequest, event.CodeInvalidRequest, err.Error())
		return
	}

	// SSE responses must keep the request-scoped context alive for the entire
	// stream; do not wrap with WithTimeout — the per-call bus options already
	// carry the same deadlines into the bus runtime.
	streamOpts := []bus.InvokeOption{}
	if a.timeout > 0 {
		streamOpts = append(streamOpts, bus.WithTimeout(a.timeout))
	}
	if a.idle > 0 {
		streamOpts = append(streamOpts, bus.WithIdleTimeout(a.idle))
	}

	s, err := a.bus.StreamInvoke(r.Context(), target, env, streamOpts...)
	if err != nil {
		writeBusError(w, err)
		return
	}
	defer s.Close()

	// SSE headers — must flush before the loop so clients see them
	// immediately even on slow first frames.
	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")
	w.Header().Set("X-Accel-Buffering", "no") // disable nginx buffering
	w.WriteHeader(http.StatusOK)

	rc := http.NewResponseController(w)
	_ = rc.Flush()

	for frame, ferr := range s.Events() {
		if ferr != nil {
			writeSSEError(w, rc, ferr)
			return
		}
		if err := writeSSEEnvelope(w, rc, frame); err != nil {
			a.log.Warn("http: sse write failed",
				"err", err,
				"target", target,
				"event_type", frame.EventType,
			)
			return
		}
	}
}

// handlePublish implements POST /v1/events/{event_type}.
func (a *Adapter) handlePublish(w http.ResponseWriter, r *http.Request) {
	ac, ok := a.authenticate(w, r)
	if !ok {
		return
	}
	eventType := r.PathValue("event_type")
	if eventType == "" {
		writeErrorJSON(w, http.StatusBadRequest, event.CodeInvalidRequest, "missing event_type")
		return
	}
	env, err := readEnvelope(r, eventType, ac)
	if err != nil {
		writeErrorJSON(w, http.StatusBadRequest, event.CodeInvalidRequest, err.Error())
		return
	}

	ctx := r.Context()
	wait := defaultPublishWait
	if a.timeout > 0 && a.timeout < wait {
		wait = a.timeout
	}
	var cancel context.CancelFunc
	ctx, cancel = context.WithTimeout(ctx, wait)
	defer cancel()

	if err := a.bus.Publish(ctx, env); err != nil {
		writeBusError(w, err)
		return
	}
	w.WriteHeader(http.StatusAccepted)
}

// writeSSEEnvelope serializes one envelope as a complete SSE event:
//
//	event: <event_type>
//	id:    <event_id>
//	data:  <envelope-json>
//
// followed by a blank line. The writer is flushed so each frame reaches the
// client immediately.
func writeSSEEnvelope(w http.ResponseWriter, rc *http.ResponseController, env *event.Envelope) error {
	body, err := json.Marshal(env)
	if err != nil {
		return err
	}
	if _, err := fmt.Fprintf(w, "event: %s\n", env.EventType); err != nil {
		return err
	}
	if env.EventID != "" {
		if _, err := fmt.Fprintf(w, "id: %s\n", env.EventID); err != nil {
			return err
		}
	}
	if _, err := fmt.Fprintf(w, "data: %s\n\n", body); err != nil {
		return err
	}
	return rc.Flush()
}

// writeSSEError emits a synthetic agent.response.error frame derived from a
// Bus-side error (idle/overall timeout, ctx cancel, decode error). Used when
// Stream.Events() yields a non-nil error.
func writeSSEError(w http.ResponseWriter, rc *http.ResponseController, srcErr error) {
	code := event.CodeAgentUnavailable
	switch {
	case errors.Is(srcErr, context.DeadlineExceeded), errors.Is(srcErr, bus.ErrIdleTimeout):
		code = event.CodeAgentTimeout
	}
	frame := event.New(event.ResponseError)
	frame.IsFinal = true
	body, _ := json.Marshal(event.ErrorPayload{
		Code:    code,
		Message: srcErr.Error(),
	})
	frame.Payload = body
	_ = writeSSEEnvelope(w, rc, frame)
}

File: sdk/python/src/openagentio/adapter/http/handlers.py
"""Route handlers mirroring pkg/adapter/http/handlers.go."""
from __future__ import annotations

import asyncio
import json
import logging
from typing import TYPE_CHECKING

from starlette.requests import Request
from starlette.responses import Response, StreamingResponse

from openagentio.bus.options import WithTimeout, WithIdleTimeout
from openagentio.bus.stream import ErrIdleTimeout
from openagentio.event.envelope import Envelope
from openagentio.event.payload import (
    CodeAgentTimeout,
    CodeAgentUnavailable,
    CodeAuthFailure,
    CodeInvalidRequest,
    ErrorPayload,
)
from openagentio.event.types import ResponseError

from openagentio.adapter.http.auth import AuthContext
from openagentio.adapter.http.envelope import read_envelope
from openagentio.adapter.http.errors import (
    write_error_json,
    write_bus_error,
    write_envelope_error,
)

if TYPE_CHECKING:
    from openagentio.adapter.http.adapter import Adapter

_DEFAULT_PUBLISH_WAIT = 5.0


async def _authenticate(adapter: Adapter, request: Request) -> tuple[AuthContext | None, Response | None]:
    """Run the configured AuthFunc. Returns (ctx, error_response)."""
    if adapter._auth is None:
        return None, None
    try:
        ac = await adapter._auth(request)
    except Exception as exc:
        return None, write_error_json(401, CodeAuthFailure, str(exc))
    return ac, None


async def handle_invoke(adapter: Adapter, request: Request) -> Response:
    ac, err = await _authenticate(adapter, request)
    if err is not None:
        return err

    target = request.path_params.get("target", "")
    if not target:
        return write_error_json(400, CodeInvalidRequest, "missing target")

    try:
        env = await read_envelope(request, "", ac)
    except ValueError as exc:
        return write_error_json(400, CodeInvalidRequest, str(exc))

    invoke_opts: list = []
    if adapter._timeout > 0:
        invoke_opts.append(WithTimeout(adapter._timeout))

    try:
        resp = await adapter._bus.invoke(target, env, *invoke_opts)
    except Exception as exc:
        return write_bus_error(exc)

    if resp.event_type == ResponseError:
        return write_envelope_error(resp)

    if not resp.payload:
        return Response(status_code=204)

    return Response(
        content=resp.payload,
        status_code=200,
        media_type="application/json",
    )


async def handle_stream(adapter: Adapter, request: Request) -> Response:
    ac, err = await _authenticate(adapter, request)
    if err is not None:
        return err

    target = request.path_params.get("target", "")
    if not target:
        return write_error_json(400, CodeInvalidRequest, "missing target")

    try:
        env = await read_envelope(request, "", ac)
    except ValueError as exc:
        return write_error_json(400, CodeInvalidRequest, str(exc))

    stream_opts: list = []
    if adapter._timeout > 0:
        stream_opts.append(WithTimeout(adapter._timeout))
    if adapter._idle > 0:
        stream_opts.append(WithIdleTimeout(adapter._idle))

    try:
        stream = await adapter._bus.stream_invoke(target, env, *stream_opts)
    except Exception as exc:
        return write_bus_error(exc)

    async def event_generator():
        try:
            async for frame in stream:
                yield _format_sse_envelope(frame)
                if frame.is_final:
                    break
        except (asyncio.TimeoutError, ErrIdleTimeout) as exc:
            yield _format_sse_error(exc)
        except Exception as exc:
            yield _format_sse_error(exc)
        finally:
            await stream.close()

    return StreamingResponse(
        event_generator(),
        status_code=200,
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )


async def handle_publish(adapter: Adapter, request: Request) -> Response:
    ac, err = await _authenticate(adapter, request)
    if err is not None:
        return err

    event_type = request.path_params.get("event_type", "")
    if not event_type:
        return write_error_json(400, CodeInvalidRequest, "missing event_type")

    try:
        env = await read_envelope(request, event_type, ac)
    except ValueError as exc:
        return write_error_json(400, CodeInvalidRequest, str(exc))

    wait = _DEFAULT_PUBLISH_WAIT
    if adapter._timeout > 0 and adapter._timeout < wait:
        wait = adapter._timeout

    try:
        await asyncio.wait_for(adapter._bus.publish(env), timeout=wait)
    except Exception as exc:
        return write_bus_error(exc)

    return Response(status_code=202)


def _format_sse_envelope(env: Envelope) -> bytes:
    body = env.to_bytes()
    parts = [f"event: {env.event_type}\n"]
    if env.event_id:
        parts.append(f"id: {env.event_id}\n")
    parts.append(f"data: {body.decode('utf-8')}\n\n")
    return "".join(parts).encode("utf-8")


def _format_sse_error(exc: BaseException) -> bytes:
    code = CodeAgentUnavailable
    if isinstance(exc, (asyncio.TimeoutError, ErrIdleTimeout)):
        code = CodeAgentTimeout
    frame = Envelope.new(ResponseError)
    frame.is_final = True
    payload = ErrorPayload(code=code, message=str(exc))
    frame.payload = json.dumps({"code": payload.code, "message": payload.message}).encode("utf-8")
    return _format_sse_envelope(frame)
